src/lib.rs

```diff
@@ -0,0 +1,181 @@
+use k8s_openapi::api::core::v1::{self as apicore, Service};
+use k8s_openapi::Resource;
+use kubewarden_policy_sdk::host_capabilities::kubernetes::{
+    list_resources_by_namespace, ListResourcesByNamespaceRequest,
+};
+use std::collections::BTreeMap;
+
+extern crate kubewarden_policy_sdk as kubewarden;
+use kubewarden::{protocol_version_guest, request::ValidationRequest, validate_settings};
+
+mod settings;
+use settings::Settings;
+
+use guest::prelude::*;
+use kubewarden_policy_sdk::wapc_guest as guest;
+
+#[no_mangle]
+pub extern "C" fn wapc_init() {
+    register_function("validate", validate);
+    register_function("validate_settings", validate_settings::<Settings>);
+    register_function("protocol_version", protocol_version_guest);
+}
+
+fn list_services(namespace: String) -> Result<Vec<Service>, String> {
+    let request = ListResourcesByNamespaceRequest {
+        api_version: "v1".to_owned(),
+        kind: Service::KIND.to_owned(),
+        namespace,
+        label_selector: None,
+        field_selector: None,
+    };
+    match list_resources_by_namespace(&request) {
+        Ok(response) => Ok(response.items),
+        Err(_) => Err("cannot list current services".to_owned()),
+    }
+}
+
+// Checks if the given services share the same set of selectors. Returns true
+// if both service the same selectors.
+fn is_service_with_duplicate_selector(service1: &Service, service2: &Service) -> bool {
+    if let Some(current_service_spec) = service1.spec.as_ref() {
+        if let Some(service_spec) = service2.spec.as_ref() {
+            // services can have empty selectors. Thus, we must check if both have selectors have
+            // something compare. Otherwise, two services with empty/none selectors can match.
+            return service_spec.selector.is_some()
+                && !service_spec
+                    .selector
+                    .as_ref()
+                    .unwrap_or(&BTreeMap::new())
+                    .is_empty()
+                && current_service_spec.selector.is_some()
+                && !current_service_spec
+                    .selector
+                    .as_ref()
+                    .unwrap_or(&BTreeMap::new())
+                    .is_empty()
```

Comment:
I think this is redundant as we already check that the selector is some above

Reply:
I have some doubts as well. I was thinking about doing something like:

```rust
if svc1.selector.is_none() && svc2.selector.is_none() {
  return it_is_fine
}

return svc1.selector == svc2.selector
```

Reply:
This change request does not cover all the posibilities. Take a look on this from the [docs](https://docs.rs/k8s-openapi/latest/k8s_openapi/api/core/v1/struct.ServiceSpec.html#structfield.selector):

```
Route service traffic to pods with label keys and values matching this selector. If empty or not present, the service is assumed to have an external process managing its endpoints, which Kubernetes will not modify. Only applies to types ClusterIP, NodePort, and LoadBalancer. Ignored if type is ExternalName. More info: https://kubernetes.io/docs/concepts/services-networking/service/
```

An empty option or an option with an empty BTreeMap should be consider in the same way. Therefore, if we only check if the option is none we miss the case where the option has empty BTreeMap.

Reply:
I see why we need to handle the case of `None` and `{}` selectors. However I didn't like the old code because I found hard to understand the flow due to all the `&&` concatenation.

I pushed fix that IMHO makes everything more readable. The code is already inside of your fork, hence it's visible right now. (Don't forget to do a pull to obtain it on your computer).

I've changed this function with commit https://github.com/kubewarden/unique-service-selector-policy/pull/1/commits/3d0dca814cf6bd93b246c69ea0a094a74d3cf805

Obviously we can keep discussing over there my proposal. We can definitely change it